packages/grid/x-data-grid-pro/src/hooks/features/columnResize/useGridColumnResize.tsx

```diff
@@ -441,10 +505,155 @@ export const useGridColumnResize = (
+        const width = paddingWidth + contentWidth;
+
+        filteredWidths.push(width);
+      }
+    }
+
+    const hasColumnMin = column.minWidth !== -Infinity && column.minWidth !== undefined;
+    const hasColumnMax = column.maxWidth !== Infinity && column.maxWidth !== undefined;
+    const current = {
+      min: hasColumnMin ? column.minWidth! : 0,
+      max: hasColumnMax ? column.maxWidth! : Infinity,
+      minContent: filteredWidths.length === 0 ? 0 : Math.min(...filteredWidths),
+      maxContent: filteredWidths.length === 0 ? 0 : Math.max(...filteredWidths),
+    };
+
+    result[column.field] = current;
+    return result;
+  }, {} as Result);
+
+  root.classList.remove(gridClasses.autosizing);
+
+  return widthsByField;
+}
+
+async function disableColumnVirtualization(
+  apiRef: React.MutableRefObject<GridPrivateApiPro>,
+  options: AutosizeOptionsRequired,
+) {
+  const state = gridColumnsStateSelector(apiRef.current.state);
+  const rowsLength = gridDataRowIdsSelector(apiRef).length;
+  const visibleColumnsLength = gridVisibleColumnFieldsSelector(apiRef).length;
+
+  const context = apiRef.current.getRenderContext();
+
+  // We render a context of at least `sampleLength` rows, centered around the current context
+  // to re-use as much as possible of the already rendered rows.
+  const halfLength = Math.floor(options.sampleLength / 2);
+  const middleRowIndex =
+    context.firstRowIndex + Math.floor((context.lastRowIndex - context.firstRowIndex) / 2);
+  const firstRowIndex = clamp(middleRowIndex - halfLength, 0, rowsLength);
+  const lastRowIndex = clamp(firstRowIndex + options.sampleLength, 0, rowsLength);
+
+  // Create our temporary render context. Note that if the new first/last row indexes are within
+  // the currently rendered range, we just reuse those instead. That case happens when the
+  // `sampleLength` is smaller than the currently rendered range.
+  const newContext = {
+    firstColumnIndex: 0,
+    lastColumnIndex: state.orderedFields.length,
+    firstRowIndex: Math.min(firstRowIndex, context.firstRowIndex),
+    lastRowIndex: Math.max(lastRowIndex, context.lastRowIndex),
+  }
+
+  apiRef.current.setColumnHeadersVirtualization(false);
+  apiRef.current.setRenderContext(newContext);
+
+  const headerContainer = apiRef.current.columnHeadersElementRef!.current!;
+  const cellsContainer = apiRef.current.virtualScrollerRef!.current!;
+  const firstRow = cellsContainer.querySelector('[role="row"]');
+
+  await Promise.all([
+    waitForDOM({
+      target: headerContainer,
+      validate: () => {
+        return headerContainer.querySelectorAll('[data-field]').length === visibleColumnsLength;
+      },
+    }),
+    firstRow
+      ? waitForDOM({
+          target: firstRow,
+          validate: () => {
+            return firstRow.querySelectorAll('[data-field]').length === visibleColumnsLength;
+          },
+        })
+      : Promise.resolve(),
+  ]);
```

Comment:
So yeah this is hacky. We need to wait until React has re-rendered the grid with virtualization turned off. But to do that, either we return promises from the virtual scroller & column headers APIs and manage their resolution with `useEffect` :( :( :(
Or we do this, which is to use a `MutationObserver` to know when the DOM has what we need.

I did a draft with the first method first but it was *UGLY* and added complexity in a lot of places. So I tried this and it feels not that ugly, despite being hacky. Comments welcome. Please share my despair.